src/main/java/com/amazon/ion/impl/bin/IonManagedWriter_1_1.kt

```diff
@@ -0,0 +1,326 @@
+        handleSymbolText(name, options.shouldWriteInline(SymbolKind.FIELD_NAME, name), userData::writeFieldName, userData::writeFieldName)
+    }
+
+    override fun setFieldNameSymbol(name: SymbolToken) {
+        handleSymbolToken(name, options.shouldWriteInline(SymbolKind.FIELD_NAME, name), userData::writeFieldName, userData::writeFieldName)
+    }
+
+    override fun addTypeAnnotation(annotation: String) {
+        if (annotation == SystemSymbols.ION_SYMBOL_TABLE && depth == 0) {
+            throw IonException("User-defined symbol tables not permitted by the managed writer.")
+        }
+        handleSymbolText(annotation, options.shouldWriteInline(SymbolKind.ANNOTATION, annotation), userData::writeAnnotations, userData::writeAnnotations)
+    }
+
+    override fun setTypeAnnotations(annotations: Array<String>?) {
+        // Interning happens in addTypeAnnotation
+        userData._private_clearAnnotations()
+        annotations?.forEach { addTypeAnnotation(it) }
+    }
+
+    override fun setTypeAnnotationSymbols(annotations: Array<SymbolToken>?) {
+        userData._private_clearAnnotations()
+        annotations?.forEachIndexed { i, it ->
+            // TODO: This is handled inconsistently. If you add annotations one at a time using addTypeAnnotation,
+            //       we don't know whether the $ion_symbol_table annotation is the first one or not.
+            if (depth == 0 && i == 0) {
+                if (it.sid == SystemSymbols.ION_SYMBOL_TABLE_SID || it.text == SystemSymbols.ION_SYMBOL_TABLE)
+                    throw IonException("User-defined symbol tables not permitted by the managed writer.")
+            }
+            handleSymbolToken(it, options.shouldWriteInline(SymbolKind.ANNOTATION, it), userData::writeAnnotations, userData::writeAnnotations)
+        }
+    }
+
+    override fun stepIn(containerType: IonType?) {
+        val newDepth = depth + 1
+        when (containerType) {
+            IonType.LIST -> userData.stepInList(options.writeDelimited(ContainerType.LIST, newDepth))
+            IonType.SEXP -> userData.stepInSExp(options.writeDelimited(ContainerType.SEXP, newDepth))
+            IonType.STRUCT -> userData.stepInStruct(options.writeDelimited(ContainerType.STRUCT, newDepth))
+            else -> throw IllegalArgumentException("Not a container type: $containerType")
+        }
+    }
+
+    override fun stepOut() = userData.stepOut()
+
+    override fun isInStruct(): Boolean = userData.isInStruct()
+
+    private inline fun <T> T?.writeMaybeNull(type: IonType, writeNotNull: (T) -> Unit) {
+        if (this == null) {
+            writeNull(type)
+        } else {
+            writeNotNull(this)
+        }
+    }
+
+    override fun writeSymbol(content: String?) {
+        if (content == null) {
+            userData.writeNull(IonType.SYMBOL)
+        } else {
+            handleSymbolText(content, options.shouldWriteInline(SymbolKind.VALUE, content), userData::writeSymbol, userData::writeSymbol)
```

Comment:
Should the inlining decision for a particular symbol be cached?

Reply:
Maybe? A HashMap lookup might be more expensive than a simple check such as `text.length > 5`.

But also, I think that caching the decision prevents some of the interesting implementation ideas that I mentioned. (E.g. only intern the symbol after it's occurred at least N times.)

Reply:
As I've thought about this more, I think that caching works well if it is implemented as part of the `SymbolInliningStrategy`. This way, the `SymbolInliningStrategy` can manage the cache invalidation in a way that works with its logic.